pkg/eventfilter/subscriptionsapi/cesql_correlationid_filter.go

```diff
@@ -0,0 +1,227 @@
+	"os"
+	"path"
+	"regexp"
+	"strings"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"knative.dev/eventing/pkg/eventfilter"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	cesql "github.com/cloudevents/sdk-go/sql/v2"
+	cefn "github.com/cloudevents/sdk-go/sql/v2/function"
+	ceruntime "github.com/cloudevents/sdk-go/sql/v2/runtime"
+)
+
+// Add a user defined function to validate correlation id, then return a cesql_filter
+func NewCESQLCorrelationIdFilter(expr string) (eventfilter.Filter, error) {
+
+	var correlationIdFilterFunction = cefn.NewFunction(
+		"KN_VERIFY_CORRELATIONID",
+		[]cesql.Type{cesql.StringType},
+		cesql.TypePtr(cesql.StringType),
+		cesql.BooleanType,
+		func(event cloudevents.Event, i []interface{}) (interface{}, error) {
+			correlationId := i[0].(string)
+
+			match, _ := regexp.MatchString(".*:.*", correlationId)
+			if !match {
+				return false, errors.New("correlationId Format: <original id>:<base64/hex encoded encrypted id>")
+			}
+
+			slice := strings.Split(correlationId, ":")
+			originalId := slice[0]
+			encryptedId := slice[1]
+
+			encryptedIdBytes, err := decodeBase64OrHex(encryptedId)
+			if err != nil {
+				return false, err
+			}
+
+			// Create a set of secret names to try looking for in k8s
+			secretNamesToTry := make(map[string]bool)
+
+			// Iterate through secret names in argument list and add them to the set
+			for num, secret := range i {
+				if num != 0 {
+					secretNamesToTry[secret.(string)] = true
+				}
+			}
+
+			secrets, err := getSecretsFromK8s()
+			if err != nil {
+				return false, err
+			}
+
+			/*
+			 * Go through each retrived secret
+			 * Check if the secret has data for a key and algorithm
+			 * Check if the secret name matches with one of the arguments
+			 * Decrypt encryptedId using the current secret's key
+			 * Check if the encrypted value matches with originalId
+			 */
+			for _, secret := range secrets {
+				key, keyFieldExists := secret.Data["key"]
+				algorithm, algorithmFieldExists := secret.Data["algorithm"]
+
+				if keyFieldExists && algorithmFieldExists && secretNamesToTry[secret.Name] {
+					var decryptionFunc func(originalId string, encryptedIdBytes []byte, key []byte) (bool, error)
+					switch strings.ToUpper(string(algorithm)) {
+					case "AES", "AES-ECB":
+						decryptionFunc = compareWithAES
+					case "DES":
+						decryptionFunc = compareWithDES
+					case "3DES", "TRIPLEDES":
+						decryptionFunc = compareWithTripleDES
+					case "RC4":
+						decryptionFunc = compareWithRC4
+					default:
+						return false, errors.New("cipher algorithm not supported")
+					}
+
+					res, err := decryptionFunc(originalId, encryptedIdBytes, key)
+					if err != nil {
+						return false, err
+					}
+					if res {
+						return true, nil
+					}
+				}
+			}
+
+			return false, nil
+		},
+	)
+
+	ceruntime.AddFunction(correlationIdFilterFunction)
```

Comment:
I might add this to the runtime somewhere else, since this will be called to make a new CESQL expression as well. The function is doing a bit too much. There should be 2 functions:
1. Make and add the CESQL function
2. Parse the CESQL expression into a filter